keps/sig-scheduling/5234-dra-resourceslice-mixins/README.md

```diff
@@ -0,0 +1,912 @@
+  //
+  // The mixins referenced here must be defined in the same
+  // ResourceSlice.
+  //
+  // The maximum number of includes is 8.
+  //
+  // +featureGate=DRAResourceSliceMixins
+  // +optional
+  // +listType=atomic
+  Includes []string
+}
+
+type DeviceCounterConsumption struct {
+  ...
+
+  // Includes defines a list of references to DeviceCounterConsumptionMixin.
+  // The counters listed in the mixins will be added to the
+  // counters that will be consumed by the device.
+  //
+  // The counters of each included mixin are applied in
+  // order. Conflicting counters from multiple mixins are taken from the
+  // last mixin listed. Counters set on the DeviceCounterConsumption will
+  // always override counters from mixins.
+  //
+  // The mixins referenced here must be defined in the same
+  // ResourceSlice.
+  //
+  // The maximum number of includes is 8.
+  //
+  // +featureGate=DRAResourceSliceMixins
+  // +optional
+  // +listType=atomic
+  Includes []string
+}
+```
+
+### Implementation
+
+The DRA scheduler will keep the mixin structure throughout the scheduling process
+as much as possible and avoid completely flattening the ResourceSlices. This is
+to avoid additional memory usage that might come as a result. For example, we
+plan to walk the mixins as part of the CEL variable lookup to avoid having to
+flatten the device representation.
+
+If the mixins feature is disabled, any devices or counter sets that
+references mixins will be droppped. This also means that all devices that references
+a dropped counter set will also be dropped. The result is that the scheduler will not
+see those devices. From the users point of view, the consequence is that the scheduler
+might pick a different device than expected or fails to allocate any device at all. But we
+think this failure mode is preferable than allowing the scheduler to make allocation decisions
+based on incomplete data.
+
+### Limits
+
+For DRA, we have gradually moved away from individual per-slice and per-map limits towards
+aggregating at the higher level. The reason for this is to give users maximum flexibility between
+defining a small number of complex devices or a large number of simple devices in a single
+ResourceSlice without exceeding the limit on the size of Kubernetes objects.
+
+For this KEP, we propose taking this to what is essentially the logical conclusion, where
+we enforce most of the limits across all devices, mixins and counter sets in a ResourceSlice,
+rather than setting separate limits for each of them.
+
+The ResourceSlice-wide limits will be:
+* Total number of devices is 128.
+* Total combined number of attributes and capacity in a ResourceSlice is 4096 (so with the maximum number of devices, there can be 32 per device).
+* Total number of counters is 256.
+* Total number of consumed counters is 2048 (so with the maximum number of devices, there can be 16 per device).
+
+We will still enforce some per-slice limits:
```

Comment:
```suggestion
We will still enforce some per-field limits:
```

My initial read expanded "per-slice" to "per-ResourceSlice", which was the opposite of what was meant...

Reply:
Yeah, agree that "per-slice" is a bit confusing. Updated.